Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -32,32 +32,92 @@ extension HTTPClient {
             ///
             /// - parameters:
             ///     - closure: function that will be called to write actual bytes to the channel.
+            @available(*, deprecated, message: "StreamWriter is deprecated, please use StreamWriter2")
             public init(closure: @escaping (IOData) -> EventLoopFuture<Void>) {
                 self.closure = closure
             }
 
+            // This is needed so we don't have build warnings in the client itself
+            init(internalClosure: @escaping (IOData) -> EventLoopFuture<Void>) {
+                self.closure = internalClosure
+            }
+
             /// Write data to server.
             ///
             /// - parameters:
             ///     - data: `IOData` to write.
+            @available(*, deprecated, message: "StreamWriter is deprecated, please use StreamWriter2")
             public func write(_ data: IOData) -> EventLoopFuture<Void> {
                 return self.closure(data)
             }
         }
 
+        public struct StreamWriter2 {
+            public let allocator: ByteBufferAllocator
+            let onChunk: (IOData) -> EventLoopFuture<Void>
+            let onComplete: EventLoopPromise<Void>
```

Comment:
Is there a reason we're writing this as a callback-taking type rather than just holding onto the task object?

Reply:
and "forward" all write requests to it's channel? that might be a good idea, will try that, thanks!